### PR TITLE
Refactor pipeline to use grain crop classes

### PR DESCRIPTION
**Pull request by SylviaWhittle**
## Main things
This PR is designed to improve how we handle grains in the processing stages of TopoStats, starting at the grain finding stage, up to the disordered tracing stage. In future, this might be extended through the disordered tracing stage and beyond, however I've restricted the scope of this PR for the sake of everyone's sanity. The reason for stopping at disordered tracing is that once disordered tracing returns, all the data is wrapped up in neatly structured dictionaries, by grain and molecule, similar to what I've implemented, so I deemed this similar enough to not bother changing it yet.

The way this PR tries to standardize how we handle grains, is using DataClasses:

- `ImageGrainCrops`
  + has two attributes, `above` and `below`, each holding a `DirectionGrainCrops` object for that direction's grain crops
- `GrainCropsDirection`
  + two attributes: `crops` and `full_mask_tensor`
  + `crops` stores dictionaries of `GrainCrop` objects (`[int, GrainCrop]`)
  + `full_mask_tensor` stores a full sized mask for the image, size is NxNxC where C is the number of classes. This is NOT automatically updated when the `crops` property is edited, this is because we don't want to update things during a loop. This can be discussed if this is an incorrect decision!
- `GrainCrop`
  + Stores various properties about the grain, such as mask, image, bbox padding etc.

This has the benefit of standardizing how we handle grains going forward, as we had previously been rather discordant in the types of data structures that we use in various parts of the codebase.

It also adds a helpful (I hope!!) layer of abstraction to processing functions, for example the `run_grainstats` function in processing no longer needs to take `image`, `grain_masks`, `pixel_to_nm_scaling`, it now takes just `image_grain_crops` which contains all the data for each crop.

This of course does come at the cost of increased memory usage as there are duplication of parts of images in the data structures as well as repeatedly listing the `pixel_to_nm_scaling` factor etc, however I personally find that the benefits here far outweigh the negatives. When working on the harbo-rings project, I found myself naturally extracting all the grains and storing them in a dictionary rather than keeping track of full image masks, I know Max also does this based on how he's handled the tracing code. 

### `disordered_tracing.py`

- Removed `prep_arrays`. Prep arrays no longer needed, since it made a dictionary of grain crops, but we already have these now with the refactor.

# TopoStats Pull Requests

Please provide a descriptive summary of the changes your Pull Request introduces.

The [Software Development](https://afm-spm.github.io/TopoStats/main/contributing.html#software-development) section of
the Contributing Guidelines may be useful if you are unfamiliar with linting, pre-commit, docstrings and testing.

**NB** - This header should be replaced with the description but please complete the below checklist or a short
description of why a particular item is not relevant.

---

Before submitting a Pull Request please check the following.

- [x] Existing tests pass.
- [x] Documentation has been updated and builds. Remember to update as required...
  - [x] `docs/configuration.md`
  - [x] `docs/usage.md`
  - [x] `docs/data_dictionary.md`
  - [x] `docs/advanced.md` and new pages it should link to.
- [x] Pre-commit checks pass.
- [x] New functions/methods have typehints and docstrings.
- [x] New functions/methods have tests which check the intended behaviour is correct.

### `topostats/default_config.yaml`

If adding options to `topostats/default_config.yaml` please ensure.

- [x] There is a comment adjacent to the option explaining what it is and the valid values.
- [x] A check is made in `topostats/validation.py` to ensure entries are valid.
- [x] Add the option to the relevant sub-parser in `topostats/entry_point.py`.

### CODE REVIEW

**Comment by SylviaWhittle** [comment]
Proposed solution to the data frame issue

```
|----------------------------------------------------------------------------------------------------------------------|
|   image   |    direction   |     class         | grain | molecule | ... <grainstats> ... | ... <dnatracingstats> ... |
| mini.spm  |    above       |   dna_only        | 0     | 0        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   dna_only        | 0     | 1        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   dna_only        | 1     | 0        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   dna_only        | 1     | 1        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   protein_only    | 0     | 0        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   protein_only    | 0     | 1        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   protein_only    | 1     | 0        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   protein_only    | 1     | 1        | ... <stats> ...      |          NONE             |
| mini.spm  |    above       |   combined_mask   | 0     | 0        | ... <stats> ...      |     ... <stats> ...       |
|----------------------------------------------------------------------------------------------------------------------|
```

**Comment by ns-rse** [comment]
Its that or split into separate files.

I'm ambivalent as to the preferred solution as I don't use the output but consideration for end users should be given. Whilst data management, manipulation, summarisation and plotting are, in my view, core skills for researchers these days experience levels vary widely and I don't know what would be easiest.

**Comment by ns-rse** [comment]
Are we aiming to include this refactoring in `v2.3.0` release?

File: topostats/processing.py [hunk 1]
```diff
@@ -430,25 +398,23 @@ def run_disordered_tracing(
...
-                if np.max(dna_class_mask) == 0:
+
+            grain_crop_direction: GrainCropsDirection
+            for direction, grain_crop_direction in image_grain_crops.__dict__.items():
+
+                if grain_crop_direction is None:
                     LOGGER.warning(
                         f"[{filename}] : No grains exist for the {direction} direction. Skipping disordered_tracing for {direction}."
                     )
                     raise ValueError(f"No grains exist for the {direction} direction")
```
Comment (MaxGamill-Sheffield) [inline topostats/processing.py:409]: Because of the presence of .above and .below every time regardless of the chosen option, the above section will always iterate through both, thus ValueError needs to be removed as we don't want to break this loop.

Then we can use:
```
if grain_crop_direction is None:
	...
else:
	trace_image_disordered(...
```

File: topostats/tracing/disordered_tracing.py [hunk 2]
```diff
@@ -297,58 +298,69 @@ def trace_image_disordered(  # pylint: disable=too-many-arguments,too-many-local
...
-    if image.shape != grains_mask.shape:
-        raise ValueError(f"Image shape ({image.shape}) and Mask shape ({grains_mask.shape}) should match.")
+    # # Check both arrays are the same shape - should this be a test instead, why should this ever occur?
+    # if image.shape != grains_mask.shape:
+    #     raise ValueError(f"Image shape ({image.shape}) and Mask shape ({grains_mask.shape}) should match.")
 
-    cropped_images, cropped_masks, bboxs = prep_arrays(image, grains_mask, pad_width)
-    n_grains = len(cropped_images)
-    img_base = np.zeros_like(image)
+    # cropped_images, cropped_masks, bboxs = prep_arrays(image, grains_mask, pad_width)
```
Comment (MaxGamill-Sheffield) [inline topostats/tracing/disordered_tracing.py:305]: (continued from default config) or need to be padded here otherwise the 3x3 convolutions will not work

File: topostats/processing.py (hunk 1 above)
Comment (ns-rse) [reply topostats/processing.py:409]: So the suggestion is simply to remove the `raise` ?

```suggestion
                    raise ValueError(f"No grains exist for the {direction} direction")
```

Comment (MaxGamill-Sheffield) [reply topostats/processing.py]: Yes but you'll also want to avoid actually trying to run the disordered tracing with an empty dict as that's not handled well I think so it needs to be in the else clause

**Comment by SylviaWhittle** [comment]
> Are we aiming to include this refactoring in `v2.3.0` release?

I don't think so necessarily. If this comes before we do `v2.3.0` then sure but don't hold it off for this PR

**Comment by SylviaWhittle** [comment]
We will need to make grain padding an option in `grains` and require at least `1`. probably want to take this out of unet config and have it general. This is because Max's code looks at adjacent pixels (so can't touch edge)

**Comment by SylviaWhittle** [comment]
Notes for self for after the break:

This PR focuses on using data classes to hold grains in a per-grain data structure rather than a per-processing-step, whole image + whole image tensor mask structure.

This PR's scope takes it from grains initially generated, to disordered tracing and no further, to keep this PR more manageable.

This PR will enable downstream processing of multi class grain tensor masks, and allow Max's code to trace them while also getting grain stats out for sub-grains, results stored neatly in disordered tracing, while avoiding any clashes of index in the grainstats csv.

In subsequent PRs, it would be good to try to align the stats for the sub-grains with sections of the molecules identified by Max's tracing code, though they may produce vastly different structures in some circumstances. Use positions potentially to do the mapping (WIP proposal).

Traditional height thresholding to produce multi class masks will also be required. This will produce "halos" but this is to be ignored / dealt with later and not worried about. Allows the lab members to at least check feasibility of their pipelines.

Pixel to nanometre scaling and filenames should be kept in the grain crops too.

**Comment by SylviaWhittle** [comment]
Things to add to grain crop objects:
- pixel to nm
- filename

File: topostats/default_config.yaml [hunk 3]
```diff
@@ -65,6 +65,7 @@ grainstats:
   extract_height_profile: true # Extract height profiles along maximum feret of molecules
 disordered_tracing:
   run: true # Options : true, false
+  class_index: 1 # The class index to trace. This is the class index of the grains.
   min_skeleton_size: 10 # Minimum number of pixels in a skeleton for it to be retained.
   pad_width: 1 # Pixels to pad grains by when tracing
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/default_config.yaml:70]: Done!

**Comment by SylviaWhittle** [comment]
How grains are handled now
image

**Comment by SylviaWhittle** [comment]
Note to self: Need to overhaul how plotting works in `run_grains`. probably want to dynamically construct the output image mask based on config in `run_grains` from `grain crops` rather than doing it in `find_grains`.

**Comment by SylviaWhittle** [comment]
Also perhaps we want to keep the original mask? Would need to possibly keep extra data / classes in the tensor in find grains to keep them?

**Comment by SylviaWhittle** [comment]
Need to return the diagnostic images from find_grains

**Comment by SylviaWhittle** [comment]
"it works on my machine!" (tests fail on GH-A, but not on my machine, seems to be due to this issue again:

image

**Comment by ns-rse** [comment]
> "it works on my machine!" (tests fail on GH-A, but not on my machine, seems to be due to this issue again:

I think this was addressed in #995.

Its down to `topoly` versions, the GitHub Action uses `topoly-1.0.4` and we can see in the changeset that there are changes in the target CSV `tests/resources/tracing/ordered_tracing/catenanes_ordered_tracing_molstats.csv` which revert to the older version.

Do you by chance have `topoly-1.0.2` on your system/virtualenv? (`pip show topoly`) If so can you upgrade to `topoly-1.0.4` and revert the above mentioned file as I think tests would then pass locally for you and on GitHub Actions.

File: topostats/grains.py [hunk 4]
```diff
@@ -1577,3 +2032,268 @@ def merge_classes(
...
+
+                # Label the regions
+                labelled_regions = Grains.label_regions(class_mask)
+                region_properties = Grains.get_region_properties(labelled_regions)
+
+                # Iterate over the regions
+                for region in region_properties:
+                    # Get the region mask
+                    region_mask = labelled_regions == region.label
+                    region_bbox = region.bbox
```
Comment (ns-rse) [inline topostats/grains.py:2228]: I think since its just accessing elements of `region.bbox` we could do that directly without losing any readability.

```suggestion
```

File: topostats/grains.py [hunk 5]
```diff
@@ -1577,3 +2032,268 @@ def merge_classes(
...
+
+                    # Check the region size
+                    if (
+                        region.area < min_object_size
+                        or (region_bbox[2] - region_bbox[0]) < min_object_bbox_size
```
Comment (ns-rse) [inline topostats/grains.py:2233]:
```suggestion
                        or (region.bbox[2] - region.bbox[0]) < min_object_bbox_size
```

File: topostats/unet_masking.py [hunk 6]
```diff
@@ -222,63 +219,94 @@ def make_bounding_box_square(
...
+    -----
+    The crop indices are inclusive on the minimum side and exclusive on the maximum side.
+    So an object like this:
+    0 0 0 0 0
+    0 1 1 1 0
+    0 1 1 1 0
+    0 1 1 1 0
+    0 0 0 0 0
+    With no padding, would have a bounding box of (1, 1, 4, 4) and not (1, 1, 3, 3).
+    Hence the maximum index for the bbox is allowed to exceed the maximum index of the image by 1.
```
Comment (ns-rse) [inline topostats/unet_masking.py:243]: I noticed in one of the tests though that a 10x10 image had a `bbox` where the max values were `12` which is more than 1, but everything works (by the looks of it!).

File: topostats/processing.py [hunk 7]
```diff
@@ -199,81 +198,140 @@ def run_grains(  # noqa: C901
                 for direction, image_arrays in grains.directions.items():
                     LOGGER.debug(f"[{filename}] : Plotting {direction} Grain Finding Images")
                     grain_out_path_direction = grain_out_path / f"{direction}"
+                    # Possibly delete this creation of the directory since we already do this earlier?
                     if plotting_config["image_set"] == "all":
-                        grain_out_path_direction.mkdir(parents=True, exist_ok=True)
```
Comment (ns-rse) [inline topostats/processing.py:203]: Not checked in detail but are we sure the `grain_out_path_direction` will exist and this line is no longer needed?

File: topostats/processing.py [hunk 8]
```diff
@@ -425,31 +458,30 @@ def run_disordered_tracing(
...
+            grain_crop_direction: GrainCropsDirection
+            for direction, grain_crop_direction in image_grain_crops.__dict__.items():
+
+                if grain_crop_direction is None:
                     LOGGER.warning(
                         f"[{filename}] : No grains exist for the {direction} direction. Skipping disordered_tracing for {direction}."
                     )
-                    raise ValueError(f"No grains exist for the {direction} direction")
-                # if grains are found
+                    continue
```
Comment (ns-rse) [inline topostats/processing.py:475]: :+1: Nice to be able to continue processing rather than erroring out.

File: topostats/grainstats.py [hunk 9]
```diff
@@ -441,8 +424,8 @@ def calculate_radius_stats(self, edges: list, points: list) -> tuple[float]:
         return {
             "min": np.min(radii),
```
Comment (ns-rse) [inline topostats/grainstats.py:425]: Curious why `np.min()` and `np.max()` don't need explicitly converting to `float()` but `np.mean()` and `np.median()` do. If anything I would have expected it to be the other way round!

File: tests/resources/tracing/nodestats/catenanes_nodestats_grainstats.csv [hunk 10]
```diff
@@ -1,3 +1,3 @@
 ,image,grain_number,num_crossings,avg_crossing_confidence,min_crossing_confidence
-grain_0,test_image,0,4,0.4013589828832889,0.2129989376767838
-grain_1,test_image,1,4,0.3441057054647598,0.17063184531586506
+grain_0,test_image,0,4,0.21426594097881008,0.001258249874731443
```
Comment (ns-rse) [inline tests/resources/tracing/nodestats/catenanes_nodestats_grainstats.csv:2]: Are we happy with these changes? They seem quite large.

**Review by ns-rse — CHANGES_REQUESTED** [review]
EPIC amount of excellent work @SylviaWhittle I think making our objects Classes is a really good decision long term and this gives us a good basis on which to build on.

I think you've seen #1092 which made some minor suggestions. I've looked through most things and think I have a general feel for things (many of the changes are names and/or where a functions options have changed removing those options).

Various in-line suggestions have been made.

File: tests/resources/tracing/nodestats/catenanes_nodestats_grainstats.csv (hunk 10 above)
Comment (SylviaWhittle) [reply tests/resources/tracing/nodestats/catenanes_nodestats_grainstats.csv:2]: I was suspicious of these changes too, however I looked at the traces and the traces are fine, though they slightly different (due to randomness Max and I believe). the differences in the traces are not significant, but they do result in some quite different results as you can see. The crossings in question for that image are very difficult and a slight change in trace does produce a large change in results, however upon manual review, we believe these changes do reflect the slightly altered (but not problematic trace)

so TLDR: yes, caused by a small change in the trace due to randomness, turns out the tracing for that image is actually slightly better now (but that's just luck)

File: tests/_regtest_outputs/test_grainstats_minicircle.test_grainstats_regression.out
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/_regtest_outputs/test_grainstats_minicircle.test_grainstats_regression.out:1]: This comment prompted me to check, and there was a bug! However I've fixed it now, and while there are still differences, I am sure the new values are correct.

Here is the evidence:

Grain stats output:
image

Image:
image


The coordinates match up very well to the visually identified centroids of the grains.

Remember that the top left is actually (0, 0) for the image.

Coords (nm) from `all_statistics.csv` -> Coords flipped in y (128 - y_coord) (nm)
y35,x87 -> y93,x87 (top right grain)
y72,x86 -> y56,x86 (bottom right grain)
y75,x38 -> y53,x38 (bottom left grain)

See that these line up perfectly with the centers of the grains.


I realise just now that this means that the `all_statistics` labels for the x and y are flipped, I'll fix that right now.

[the remark this answers is not in the document]
Comment (ns-rse) [reply tests/_regtest_outputs/test_grainstats_minicircle.test_grainstats_regression.out:1]: Good work spotting that :+1:

File: topostats/default_config.yaml [hunk 11]
```diff
@@ -66,8 +66,8 @@ grainstats:
   extract_height_profile: true # Extract height profiles along maximum feret of molecules
 disordered_tracing:
   run: true # Options : true, false
+  class_index: 1 # The class index to trace. This is the class index of the grains.
   min_skeleton_size: 10 # Minimum number of pixels in a skeleton for it to be retained.
-  pad_width: 1 # Pixels to pad grains by when tracing
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/default_config.yaml:70]: So the way this works is that the `GrainCrop` objects that are passed into `disordered_tracing` are pre-cropped so don't need to be cropped again during `disordered_tracing` and subsequent steps, so we don't configure a padding for them since it's already done - if that makes sense?

File: tests/_regtest_outputs/test_processing.test_process_scan_above.out
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/_regtest_outputs/test_processing.test_process_scan_above.out:1]: same as resolved issue above 👍

File: topostats/grains.py [hunk 12]
```diff
@@ -35,11 +36,512 @@
...
+        return (
+            np.array_equal(self.image, other.image)
+            and np.array_equal(self.mask, other.mask)
+            and self.padding == other.padding
+            and self.bbox == other.bbox
+            and self.pixel_to_nm_scaling == other.pixel_to_nm_scaling
+            and self.filename == other.filename
+        )
+
+    def debug_locate_difference(self, other: object) -> None:
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/grains.py:325]: complete necessity! 😛

File: tests/conftest.py [hunk 13]
```diff
@@ -325,6 +326,59 @@ def random_grains(grains_config: dict, random_filters: Filters) -> Grains:
...
+                ]
+            ),
+        ],
+        axis=-1,
+    )
+    return GrainCrop(
+        image=image,
+        mask=mask,
+        padding=2,
+        bbox=(1, 1, 12, 12),
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/conftest.py]: fixed, should have been 11, 11, thanks!

File: docs/advanced/grain_finding.md [hunk 14]
```diff
@@ -64,13 +64,43 @@ maximum area of the grains that you want to keep, in nanometers squared. Eg if y
...
+
+Initially, this tensor is just given two binary classes, `0`, and `1`. Class `0` is the background class and is 1 where
+there are no grains in the other classes and 0 where there are grains in any of the other classes. Class `1` is simply
+the mask produced by the basic grain detection, which usually is designed to be segmenting `DNA` but could really be
+anything.
+
+Multi class masking is able to be performed on the existing grains by using a U-Net (a deep-learning feature detection
+program) which can optionally be ran after simple grain finding, though needs training.
+
+It is planned to be able to run multiple rounds of traditional thresholding instead of running a U-Net for easier but
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply docs/advanced/grain_finding.md:85]: Done 👍

File: topostats/grains.py [hunk 15]
```diff
@@ -1433,25 +1897,22 @@ def convert_classes_when_too_big_or_small(
...
         class_size_thresholds: list[tuple[int, int, int]] | None,
         class_region_number_thresholds: list[tuple[int, int, int]] | None,
         nearby_conversion_classes_to_convert: list[tuple[int, int]] | None,
         class_touching_threshold: int,
         keep_largest_labelled_regions_classes: list[int] | None,
         class_connection_point_thresholds: list[tuple[tuple[int, int], tuple[int, int]]] | None,
-    ) -> npt.NDArray:
+    ) -> dict[int, GrainCrop]:
         """
         Vet grains in a grain mask tensor based on a variety of criteria.
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply topostats/grains.py:1910]: Made and issue 👍

File: topostats/unet_masking.py (hunk 6 above)
Comment (SylviaWhittle) [reply topostats/unet_masking.py:243]: Fixed I think, though there are currently no checks on the fly to ensure that the bbox is correct since the only time it would be wrong is if we manually set it to be wrong (like when constructing tests!)

Could add an getter-setter check later on if this becomes an issue though 😄 (personally love getter-setter sanity checking)

File: topostats/grainstats.py (hunk 9 above)
Comment (SylviaWhittle) [reply topostats/grainstats.py:425]: I believe so? Here there is code in `processing.py` > `get_out_paths` that does it:

image


Which is ran as the first step in `processing.py` > `process_scan`:
image


Might be understanding this wrong though ^^

File: tests/test_entry_point.py [hunk 16]
```diff
@@ -283,6 +285,7 @@ def test_entry_point_subprocess_help(capsys, argument: str, option: str) -> None
             run_modules.grains,
             {
                 "config_file": Path("dummy/config/dir/config.yaml"),
+                "grain_crop_padding": 1,
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/test_entry_point.py:288]: It changes how much of a gap there is between the edge of the grain and the bounding box. Higher values enable better U-Net performance **in some situations** and also there must be at least 1 pixel gap to allow `disordered_tracing` to run.

I am happy with this default and the whole lab has tested their data and no issues have been found relating to the padding IIRC

[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/test_entry_point.py:288]: Also means we don't need to have the `pad_width` option in the tracing steps as it's already been cropped as opposed to having to be cropped later on from the full grain mask in each step

File: tests/resources/tracing/ordered_tracing/catenanes_ordered_tracing_grainstats.csv [hunk 17]
```diff
@@ -1,3 +1,3 @@
 ,image,grain_number,num_mols,writhe_string
 grain_0,catenane,0,2,++++
-grain_1,catenane,1,2,+-++
+grain_1,catenane,1,2,++++
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/resources/tracing/ordered_tracing/catenanes_ordered_tracing_grainstats.csv:3]: image
👍 from Max <from slack>

File: tests/resources/tracing/ordered_tracing/catenanes_ordered_tracing_molstats.csv [hunk 18]
```diff
@@ -1,5 +1,5 @@
 ,image,grain_number,molecule_number,circular,topology,topology_flip,processing
 0,catenane,0,0,True,L4a1,L2a1,nodestats
 1,catenane,0,1,True,L4a1,L2a1,nodestats
-2,catenane,1,0,True,L2a1,0_1U0_1,nodestats
-3,catenane,1,1,True,L2a1,0_1U0_1,nodestats
+2,catenane,1,0,True,L4a1,L2a1,nodestats
```
[the remark this answers is not in the document]
Comment (SylviaWhittle) [reply tests/resources/tracing/ordered_tracing/catenanes_ordered_tracing_molstats.csv:4]: I'm not sure but Max says it's okay 😅

File: topostats/grains.py [hunk 19]
```diff
@@ -35,11 +36,512 @@
...
+        ----------
+        value : npt.NDArray
+            Numpy array of the mask.
+
+        Raises
+        ------
+        ValueError
+            If the mask dimensions do not match the image.
+        """
+        if value.shape[0] != self.image.shape[0] or value.shape[1] != self.image.shape[1]:
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply topostats/grains.py:157]: Ah ok, thanks, hadn't thought that through properly (despite having read the tensor added a dimension to the array!).

**Review by ns-rse — APPROVED** [review]
